### PR TITLE
Improve SitesNeighbourhoods selector

### LINKED ISSUE #885: Change the Site's edit form to properly display and allow selection of secondary neighbourhoods in a more sensible way

### PR DESCRIPTION
**Pull request by alexandria-gfsc**
Improves Secondary Neighbourhood selection in Sites Admin view
Fixes #885 

- Changed sites/_form.html.erb to use Cocoon
  - Added sites/_sites_neighbourhood_fields.html.erb subform
  - Added helpers/sites_helper.rb for Sites Neighbourhood collection
  - Added @nathvda/cocoon
  - Added behaviours.site.js (for select2 single option select)
  - Added contextual_name() in Neighbourhoods model
  - Added name() in SitesNeighbourhood model
- Migrate Neighbourhoods to memoize parent.name to avoid complex lookup

### CODE REVIEW

File: app/views/admin/sites/_sites_neighbourhood_fields.erb [hunk 1]
```diff
@@ -0,0 +1,2 @@
+<%= f.input :name %>
+<%= f.hidden_field :relation_type, value: "Secondary" %>
```
Comment (alexandria-gfsc) [inline app/views/admin/sites/_sites_neighbourhood_fields.erb:2]: lol ignore this file. didn't realise i hadn't deleted it. it was a misnamed file

File: test/integration/admin/sites_integration_test.rb [hunk 2]
```diff
@@ -36,8 +36,8 @@ class AdminSitesIntegrationTest < ActionDispatch::IntegrationTest
     assert_select 'label', 'Hero image credit'
 
     # See all neighbourhoods
-    assert_select '.site__neighbourhoods' do
-      assert_select 'label', @number_of_neighbourhoods
+    assert_select '#site_neighbourhood_ids' do
+      assert_select 'option', @number_of_neighbourhoods
```
Comment (alexandria-gfsc) [inline test/integration/admin/sites_integration_test.rb:40]: this testing code is from prior commits. disregard

File: app/javascript/src/behaviors/behaviors.site.js [hunk 3]
```diff
@@ -0,0 +1,11 @@
+jQuery.extend(Behaviors, {
+  site: {
+    form: {
+      init: function() {
+        $( ".select2" ).select2({
+          multiple: false
```
Comment (kimadactyl) [inline app/javascript/src/behaviors/behaviors.site.js:6]: You can use this to pull in fuzzy search results via ajax, check docs :)

File: app/views/admin/sites/_sites_neighbourhood_fields.html.erb [hunk 4]
```diff
@@ -0,0 +1,7 @@
+<div class="nested-fields">
+    <%= f.input :name, collection: options_for_sites_neighbourhoods, include_blank: false,
+                input_html: { class: 'form-control select2' } %>
+    <%= f.hidden_field :relation_type, value: 'Secondary' %>
+    <%= link_to_remove_association 'Remove Secondary Neighbourhood', f, class: "btn btn-danger" %>
```
Comment (kimadactyl) [inline app/views/admin/sites/_sites_neighbourhood_fields.html.erb:5]: let's tidy these up a bit with buttons to right of ward names, prob need to check bootstrap docs

File: app/views/admin/sites/_form.html.erb [hunk 5]
```diff
@@ -61,22 +61,14 @@
...
-          <% end %>
-          <%= neighbourhood.name %>
-        </label><br>
-      <% end %>
+  <div class="sites_neighbourhoods">
+    <%= f.simple_fields_for :sites_neighbourhoods do |neighbourhood| %>
+      <%= render 'sites_neighbourhood_fields', :f => neighbourhood %>
     <% end %>
+    <div class="links">
+      <%= link_to_add_association 'Add secondary', f, :sites_neighbourhoods, class: "btn btn-primary" %>
```
Comment (kimadactyl) [inline app/views/admin/sites/_form.html.erb:69]: button text: Add Secondary Neighbourhood

File: app/javascript/packs/admin.js [hunk 6]
```diff
@@ -1,4 +1,4 @@
-require("cocoon")
+require("@nathanvda/cocoon")
```
Comment (kimadactyl) [inline app/javascript/packs/admin.js:1]: alphabatize these while we're here

**Review by kimadactyl — CHANGES_REQUESTED** [review]
See comments

**Comment by kimadactyl** [comment]
Is this loading neighbourhoods via ajax?

**Review by kimadactyl — CHANGES_REQUESTED** [review]
- [ ] Make fuzzy search type input work on new selects
- [ ] All selects currently set to Abbey, Leics
- [ ] Make delete link just a link and tidy up UX a lil. i would put all of the inputs together in a fieldset

**Comment by alexandria-gfsc** [comment]
Rebased on main

**Review by kimadactyl — CHANGES_REQUESTED** [review]
- [x] A few UX suggetions as per discord
- [x] Add better description text under headings
- [x] Fix create new site picker for main neighbourhood

**Comment by alexandria-gfsc** [comment]
> * [x]  A few UX suggetions as per discord
> * [x]  Add better description text under headings

Done :)

> * [x]  Fix create new site picker for main neighbourhood

hmm it is not broken for me, it just doesn't display as a select2 element for whatever reason. Moved it to a `f.input` element so the rendering is fixed

**Review by kimadactyl — CHANGES_REQUESTED** [review]
Suggested controller tests:

- [ ] Create new site with primary neighbourhood only saves. Ditto for edit/update
- [ ] Create new site with both primary and secondary neighbourhood saves. Ditto for edit/update.
- [ ] Create new site with primary neighbourhood == secondary neighbourhood just saves the primary neighbourhood. Ditto for edit/update
- [ ] Same as above but with a few secondary neighbourhoods
- [ ] Primary neighbourhood is mandatory (might exist already)

**Comment by kimadactyl** [comment]
Merged, controller tests created as another ticket

**Pull request merged**